### PR TITLE
ExPlat: Self-Contained Client Internals - Fixes

### PR DESCRIPTION
**Pull request by jessie-ross**
This PR improves the internal functions and tests before I PR the unit tests for the API:

* Remove `isRecent` and it's related checks.
* Have state be stored externally - this is important for testing as it allows client instances to have separate state.
* Fix `fallbackExperiment` name bug - it must always have a valid name.
* Simplify `asyncOneAtATime` as per @withinboredom's suggestions
* Add unit tests for rejections for Timing functions.

Part of a series rebuilding the reference WIP PR #48475

#### Testing instructions

* Automated testing

### CODE REVIEW

File: packages/explat-client/src/internal/timing.ts
Comment (yanirs) [inline packages/explat-client/src/internal/timing.ts:46]: :tada:

File: packages/explat-client/src/internal/state.ts
Comment (yanirs) [inline packages/explat-client/src/internal/state.ts]: Can please explain the need for this more? If it's meant to be passed in by clients, shouldn't it be a part of the config?

Comment (yanirs) [reply packages/explat-client/src/internal/state.ts]: Ah, I think I see what's going on given #49832. Is it so that each client would have its own store, but the store is still hidden from the client?

If we're going with this, I think it'd be cleaner for this to be more of a class-like interface, as the `store` is mutable. That is, create a store object with with methods `storeExperimentAssignment()` and `retrieveExperimentAssignment()`.

Comment (jessie-ross) [reply packages/explat-client/src/internal/state.ts]: It is used like this: 
https://github.com/Automattic/wp-calypso/pull/48475/files#diff-6ccc8c485f7dc1c524767b67ae2efa00edf70f04ab35c903a9f0287b644d507aR63

Not meant to be passed in by clients.

Comment (jessie-ross) [reply packages/explat-client/src/internal/state.ts]: Oops I didn't see your last comment: Yes we could be using a class here (and also for the ExPlat client), but I think this is simplest, will discourage adding extra state, and makes unit testing a breeze.

Comment (yanirs) [reply packages/explat-client/src/internal/state.ts]: I disagree about it being simplest, especially when looking at the way it's used. I really don't understand the aversion to using classes when they make sense...

`store.storeExperimentAssignment( experimentAssignment )` is clearer to me than `Store.storeExperimentAssignment( store, experimentAssignment )`, as it's more self-documenting that it operates on the `store`, which is a mutable object. It's also in line with browser APIs like `localStorage`.

Comment (jessie-ross) [reply packages/explat-client/src/internal/state.ts]: > I really don't understand the aversion to using classes when they make sense...

The way I see it is that classes are inherently more complex than functions - they add another layer of modelling and most importantly developers are more likely to misuse them (adding state, adding methods, mixing up purposes because it is easier to get at the state from within, needing to add more methods to utilise private methods). 

If we aren't using features that classes provide over functions like inheritance, or run-time polymorphism and we aren't using it often and everywhere, to me the added "convenience" of using them doesn't outweigh the added complexity.

So as a general rule, where they are providing the same features except for less argument to functions, I like to avoid classes and prefer functions.

> store.storeExperimentAssignment( experimentAssignment ) is clearer to me than Store.storeExperimentAssignment( store, experimentAssignment ), as it's more self-documenting that it operates on the store, which is a mutable object.

It is true we are passing a mutable state reference in, but I see this as a pretty small difference, I don't think any Javascript dev is going to trip over it.

> It's also in line with browser APIs like localStorage.

Browser APIs like localStorage are often the result of years of legacy, compromise and committees. We should be working from modern principles and practices in Javascript. 

It is also a pretty simple tool that isn't to be used externally so familiarity shouldn't be a problem.

Comment (yanirs) [reply packages/explat-client/src/internal/state.ts]: I'm sorry, but I don't buy it. What you're framing as "modern principles and practices in JavaScript" in this case looks a lot like abstract data types I saw implemented in C in my computer science degree, before they introduced us to C++. And I won't even get into the JS world's tendency to reinvent the wheel every month with a shiny new framework. :trollface: 

But anyway, focusing on the matter at hand, it's worth looking at what the implementation would look like:

```typescript
class Store {
  private experimentNameToExperimentAssignment: Record< string, ExperimentAssignment | undefined > = {}

  public storeExperimentAssignment( experimentAssignment: ExperimentAssignment ): void {
	  Validations.validateExperimentAssignment( experimentAssignment );
    const previousExperimentAssignment =
      this.experimentNameToExperimentAssignment[ experimentAssignment.experimentName ];
    if (
      previousExperimentAssignment &&
      experimentAssignment.retrievedTimestamp < previousExperimentAssignment.retrievedTimestamp
    ) {
      throw new Error(
        'Trying to store an older experiment assignment than is present in the store, likely a race condition.'
      );
    }

    this.experimentNameToExperimentAssignment[ experimentAssignment.experimentName ] = experimentAssignment;
  }

  public retrieveExperimentAssignment( experimentName: string ): ExperimentAssignment | undefined {
    return this.experimentNameToExperimentAssignment[ experimentName ];
  }
}
```

This is more than "less arguments to functions". It:

* Self-documents the mutability of the underlying object.
* Obviates the need for a custom function like `createStore()` &ndash; we get the constructor for "free" from TypeScript.
* Clearly links the store and retrieve method to each other and to the Store object they operate on.

I don't see why we'd avoid using this implementation here, especially as developer misuse is unlikely to be an issue. It seems like a perfect use case for a simple class to me.

Comment (jessie-ross) [reply packages/explat-client/src/internal/state.ts]: > I'm sorry, but I don't buy it. What you're framing as "modern principles and practices in JavaScript" in this case looks a lot like abstract data types I saw implemented in C in my computer science degree, before they introduced us to C++. And I won't even get into the JS world's tendency to reinvent the wheel every month with a shiny new framework. :trollface:

This was in reply to your suggestion of basing it off of the `localStorage` browser API, rather I believe I am providing you with a first principles based argument and I think we should stick to that.

> I don't see why we'd avoid using this implementation here, especially as developer misuse is unlikely to be an issue. It seems like a perfect use case for a simple class to me.

It seems to me that all of the points you list (except for "getting the constructor for free" which I think doesn't have much material benefit) are in-fact for the purpose of making it harder for developers to misuse the code: self-documenting the mutability, linking the functions together and to the data. So to me it looks like this discussion does come down to whether classes or functions prevent developer misuse more.

I am happy to go with your call but I still believe that using classes in this kind of case creates more issues than they solve - including the benefits you mention.

Comment (yanirs) [reply packages/explat-client/src/internal/state.ts]: > This was in reply to your suggestion of basing it off of the `localStorage` browser API, rather I believe I am providing you with a first principles based argument and I think we should stick to that.

Yeah, I disagree that the "first principles" you cite are indeed universally accepted, which means that they're not first principles (happy to be proven wrong, though). I think that the `localStorage` browser API (or more generally, [the `Storage` API it implements](https://developer.mozilla.org/en-US/docs/Web/API/Storage)) would be implemented pretty similarly these days, as it's a fairly natural way of modelling it in languages that support classes. As noted, having a collection of functions operating on an opaque data structure reminds me more of languages that don't support classes, like C.

> So to me it looks like this discussion does come down to whether classes or functions prevent developer misuse more.

I don't see it as misuse &ndash; it's more a matter of clarity.

> I am happy to go with your call but I still believe that using classes in this kind of case creates more issues than they solve - including the benefits you mention.

Yeah, it's a matter of opinion to an extent, but it feels a bit like just a general aversion to using classes based on the potential future changes to the code, which I don't think is supported by much concrete evidence. Future misuse and messiness are possible with any approach, so it's more about the current readability of the code to me.

Let's get another opinion from @withinboredom on the readability of either solution and go with that.

Comment (withinboredom) [reply packages/explat-client/src/internal/state.ts]: Holy discussion Batman!

> I really don't understand the aversion to using classes when they make sense...

There are [no such things as classes in older Javascript](https://www.typescriptlang.org/play?target=1#code/MYGwhgzhAEDKAuB7ATgU2gbwFDWgB2QEsA3MedVADz1SIFtUA7eAOTAYBVEBRa2whswCCUQgHNGg+AC5oAJVTAUAEwA80CPCKMxAGmhhGAT2gAfaAFdGy1ADNCjVMugA+aAF5MAXyw58FgCMQQmANJDReGnomeBEIcUkYgApoKiiBGLiEqVlDEwBKWWJEQmdsAEhcAHoqgHcHZURagDpSYOUyVEj+KSyJKRS0nszRfpjofIBuP1wlRk18NGJCRAsIbujhUcTmDxncaHgAC0IIZqHN1nZULg2Mrfix5gBtVL5Lvp34c-f7q4ZoABdaYHaCEWzQJL7XAEVDLVbrX69bZSaAAMjR0Le6WRjy+zTQWkIcKcHAymnYeGg6lh8LWd1x2RiBNQRJJyjJDApdDw+3ymCxx2QTWgjlq0G4yGFyChoNBAHIOMgjA4xIdEGEUOhDNBECAbMhscNdpA8ajjjrTotUBBxg5Dkd0JotfpggBrVAgExgaDIMDAdBzZSEeArRjNeVYqb7Hz7Y6nH44mJsTg8JEjM0xV4XP6fKSJ43-dCAjxGj4omIg6CxmGBYKhQlEEkMjNM5iDdPMFOoWSabRqwoGYxmSzWOwOJwC0GEizIRgOhM5qTd26d2IVl5lv7doFVnxeIA) and in newer browsers that support the `class` keyword, [it's really just syntactical sugar](https://medium.com/@naveenkarippai/debunking-fake-classes-in-javascript-78f67a6b5c96#:~:text=Dec%2012%2C%202016%C2%B77%20min%20read&text=TL%3BDR%3A%20ECMAScript6%20introduced%20class,compared%20to%20complex%20Function%20Constructors.). Javascript is a prototypical language and doesn't have inheritance like in OOP based languages. The `class` keyword in TypeScript allows one to code as though it does by transpiling any specific logic needed for it.

Using classes may add significant complexity to the code that is actually executed by the browser and slow things down. They should be used sparingly unless the use-case calls for it.

> Let's get another opinion from @withinboredom on the readability of either solution and go with that.

FWIW, I think we should go with a non-class solution for storing the state. If a developer desires, they can store the object in `localstorage` if they want to, or send it over the wire somewhere else. If it's a class with behavior, it makes things more complicated when loading it from `localstorage`, `IndexedDb`, or wherever.

This all looks like idiomatic Javascript to me.

File: packages/explat-client/src/internal/experiment-assignments.ts
Comment (withinboredom) [inline packages/explat-client/src/internal/experiment-assignments.ts]:
```suggestion
): ExperimentAssignment => ({
		experimentName: isName( experimentName ) ? experimentName : 'fallback_experiment_assignment',
		variationName: null,
		retrievedTimestamp: Timing.monotonicNow(),
		ttl: fallbackExperimentAssignmentTtl,
		isFallbackExperimentAssignment: true,
	});
};
```

File: packages/explat-client/src/internal/state.ts
Comment (jessie-ross) [reply packages/explat-client/src/internal/state.ts]: > Yeah, I disagree that the "first principles" you cite are indeed universally accepted, which means that they're not first principles (happy to be proven wrong, though).

I mean first principles as an approach to arguments based on deductions and assumptions, it isn't necessarily about agreeing on assumptions. The main body of [my reply here](https://github.com/Automattic/wp-calypso/pull/49829#discussion_r572616530) rather than the rebuttal of your point. 

> I think that the localStorage browser API (or more generally, the Storage API it implements) would be implemented pretty similarly these days, as it's a fairly natural way of modelling it in languages that support classes. As noted, having a collection of functions operating on an opaque data structure reminds me more of languages that don't support classes, like C.

If it is done there it doesn't necessarily make it right here, and just because using data structures in C isn't great that doesn't mean it applies here. (And lots of people still use C like that, particularly in the open source world.)

> I don't see it as misuse – it's more a matter of clarity.

For what purpose is clarity useful if not for aiding future development?

> it feels a bit like just a general aversion to using classes based on the potential future changes to the code, which I don't think is supported by much concrete evidence.

I feel like I have given pretty strong reasons not to use classes in these trivial cases - complexity from modelling and problems of developer misuse outweighing the benefits of self-documentation and clarity. Given that there is no concrete evidence on either side I don't see anything wrong with preferring functions over classes for trivial cases.

> Let's get another opinion from @withinboredom on the readability of either solution and go with that.

Why is readability the target and not maintainability?

File: packages/explat-client/src/internal/experiment-assignments.ts
Comment (withinboredom) [reply packages/explat-client/src/internal/experiment-assignments.ts]: It would also benefit the reader to put these in some kind of order (perhaps by length?). I found playing "where's Waldo" with the `:` fairly annoying. Not the end of the world though.

File: packages/explat-client/src/internal/test/state.ts
Comment (withinboredom) [inline packages/explat-client/src/internal/test/state.ts]: Can we pick a better name than `State`? That name seems rather ambiguous, it may be better to use the name `ExperimentState` or something. I suck at naming things, so take it with a grain of salt.

File: packages/explat-client/src/internal/test/state.ts
Comment (withinboredom) [inline packages/explat-client/src/internal/test/state.ts]: I'm not sure what benefits we get from passing in `store` everywhere and we probably don't want experimenters just storing any random variation, bypassing the API. It smells like it might cause an issue if someone deploys that by accident, or even worse, they deploy it and it only gets called under certain conditions, thereby breaking their experiment.

File: packages/explat-client/src/internal/state.ts
Comment (withinboredom) [inline packages/explat-client/src/internal/state.ts]: I went on a walk in the snow. I really don't see the benefit of this abstraction. Can you expand on this more or provide examples of how it is better and why?

File: packages/explat-client/src/internal/test/state.ts
Comment (jessie-ross) [reply packages/explat-client/src/internal/test/state.ts]: The store is local to a ExPlatClient and can't be accessed from outside of it.
These functions are also within a `src/internal` and not meant to be used outside of it.

(This touches on the Class vs Function argument btw)

File: packages/explat-client/src/internal/state.ts
Comment (jessie-ross) [reply packages/explat-client/src/internal/state.ts]: If you mean having an object around it, yeah not entirely necessary but does as a guide for adding further state - I like to keep everything together.

I'm happy to remove it.

Comment (withinboredom) [reply packages/explat-client/src/internal/state.ts]: > as a guide for adding further state

Makes sense. As someone who will happily write a single case switch statement to "hint" to people that they should write their new case there... people won't do that, in my experience; they'll write an if-statement right above the switch :sigh:. It's probably better to wait until we have more state before wrapping it in an object.

File: packages/explat-client/src/internal/test/state.ts
Comment (withinboredom) [reply packages/explat-client/src/internal/test/state.ts]: > This touches on the Class vs Function argument btw

Yeah, but arguing over implementation details/preferences is not worth it IMHO. I just want to understand the need for this abstraction when it didn't seem to change much in the tests. If it needs to exist or makes things better, then we should keep it.

> The store is local to a ExPlatClient and can't be accessed from outside of it.
> These functions are also within a src/internal and not meant to be used outside of it.

I'm still wary about it. People will use whatever is necessary to accomplish a task, especially if there's a time crunch and/or they don't understand the implications.

Comment (jessie-ross) [reply packages/explat-client/src/internal/test/state.ts]: > If it needs to exist or makes things better, then we should keep it.

Oh you can see it in action in the next PR: https://github.com/Automattic/wp-calypso/pull/48475/files#diff-6ccc8c485f7dc1c524767b67ae2efa00edf70f04ab35c903a9f0287b644d507aR63

Comment (withinboredom) [reply packages/explat-client/src/internal/test/state.ts]: > Oh you can see it in action in the next PR

Ok, here's my feedback getting a better picture of how this is all laid out.

1. The `state` object passed to the client is a leaky-abstraction. The client should never (be allowed to) manipulate the state directly if we're going to encapsulate it here.
2. From a client's perspective, the `State` object is more like a key to access other functions than something useful. Perhaps we should get rid of the `state` object we return, and instead, just use `Symbol.for( experimentName )` as the key.
3. Each time we create a client, it appears to be an entirely new client with new state. This has some interesting implications if not wielded correctly. I could also be reading it wrong. We should have some stable-state, perhaps keyed by Symbol, so we don't have nested components each creating their own clients and doing their own API calls.

File: packages/explat-client/src/internal/state.ts
Comment (yanirs) [reply packages/explat-client/src/internal/state.ts]: > If it is done there it doesn't necessarily make it right here, and just because using data structures in C isn't great that doesn't mean it applies here. (And lots of people still use C like that, particularly in the open source world.)

That's missing my point. C's syntax is limited. In languages that have richer syntax, its worth using it to simplify things.

> Why is readability the target and not maintainability?

Readability and maintainability are pretty synonymous.

> For what purpose is clarity useful if not for aiding future development?

Yes, clarity is useful for aiding future development. I just don't buy the dangers of "misuse", which are avoided by owning and reviewing code.

> I feel like I have given pretty strong reasons not to use classes in these trivial cases - complexity from modelling and problems of developer misuse outweighing the benefits of self-documentation and clarity. Given that there is no concrete evidence on either side I don't see anything wrong with preferring functions over classes for trivial cases.

Well, we can agree to disagree on this one, and go with @withinboredom's opinion as the tie breaker (i.e., stick with the current implementation). :slightly_smiling_face:

File: packages/explat-client/src/internal/test/state.ts
Comment (yanirs) [reply packages/explat-client/src/internal/test/state.ts]: Maybe call it `Store` or `AssignmentStorage` or `AssignmentStore`?

File: packages/explat-client/src/internal/state.ts
Comment (yanirs) [reply packages/explat-client/src/internal/state.ts]: It'd all be pretty self-documenting with a simple class, but I'm done arguing on this PR. :slightly_smiling_face:

File: packages/explat-client/src/internal/test/state.ts
Comment (jessie-ross) [reply packages/explat-client/src/internal/test/state.ts]: > The state object passed to the client is a leaky-abstraction. The client should never (be allowed to) manipulate the state directly if we're going to encapsulate it here.

Ah, good point about encapsulation being necessary here, this does tip the balance into class territory for me.

> Each time we create a client, it appears to be an entirely new client with new state. This has some interesting implications if not wielded correctly. I could also be reading it wrong. We should have some stable-state, perhaps keyed by Symbol, so we don't have nested components each creating their own clients and doing their own API calls.

Since we are doing single fetch experiments, the only thing that I can see go wrong is a single experiment using two separate libs, with slightly different timing giving half implemented variations. I think when we get more clients within the same codebase we can add in a dev-mode error for multiple clients

File: packages/explat-client/src/internal/test/state.ts
Comment (jessie-ross) [reply packages/explat-client/src/internal/test/state.ts]: Going with `ExperimentAssignmentStore` for the class.

File: packages/explat-client/src/internal/experiment-assignments.ts
Comment (jessie-ross) [reply packages/explat-client/src/internal/experiment-assignments.ts]: > It would also benefit the reader to put these in some kind of order (perhaps by length?). I found playing "where's Waldo" with the : fairly annoying. Not the end of the world though.

Ah I do use this order on purpose and everywhere (I hope) - ordered by "importance". Experiment Name, then Variation Name, then retrieved timestamp, then ttl, then (optionally) fallback status. This order will be kept when printing/debugging/developing (in typescript aware editors).

File: packages/explat-client/src/internal/experiment-assignment-store.ts
Comment (yanirs) [inline packages/explat-client/src/internal/experiment-assignment-store.ts]: Shall we make this `private`?

Also, `Assigntments` has a typo, but it may be better to go for a shorter name for the mapping like `experimentNameToAssignment`.

File: packages/explat-client/src/internal/experiment-assignment-store.ts
Comment (yanirs) [inline packages/explat-client/src/internal/experiment-assignment-store.ts]: "hold ExperimentAssignment state" isn't very descriptive. Maybe "store existing ExperimentAssignments in memory" or something similar?

File: packages/explat-client/src/internal/test/experiment-assignment-store.ts
Comment (yanirs) [inline packages/explat-client/src/internal/test/experiment-assignment-store.ts]: It'd be good to also add a test for trying to store an invalid experiment assignment.

File: packages/explat-client/src/internal/experiment-assignments.ts
Comment (yanirs) [inline packages/explat-client/src/internal/experiment-assignments.ts]: Under what circumstances would we call this with an invalid experiment name?

Comment (yanirs) [reply packages/explat-client/src/internal/experiment-assignments.ts]: Ah, I see it in #49832. I think it's problematic because it's a surprising result to get an assignment with a different name from what the calling code requested. 

By the way, this is the sort of change that I think would make more sense as part of #49832 (i.e., where it's needed), as trying to keep the context in mind across PRs over multiple weeks is pretty hard. It'd make it easier to review if we tried to minimise the number of open PRs by deploying often.

Comment (jessie-ross) [reply packages/explat-client/src/internal/experiment-assignments.ts]: > By the way, this is the sort of change that I think would make more sense as part of #49832 (i.e., where it's needed), as trying to keep the context in mind across PRs over multiple weeks is pretty hard. It'd make it easier to review if we tried to minimise the number of open PRs by deploying often.

This PR is almost entirely changes that are needed for #49832, split so that it won't be so big.

> Ah, I see it in #49832. I think it's problematic because it's a surprising result to get an assignment with a different name from what the calling code requested.

Yeah so this is necessary for experiments with invalid experiment names, we want to always have a valid ExperimentAssignment. I could add an `originalExperimentName` to keep the original name around.

File: packages/explat-client/src/internal/experiment-assignments.ts
Comment (jessie-ross) [inline packages/explat-client/src/internal/experiment-assignments.ts]: > Yeah so this is necessary for experiments with invalid experiment names, we want to always have a valid ExperimentAssignment. I could add an originalExperimentName to keep the original name around.

I ended up implementing it.

File: packages/explat-client/src/types.ts
Comment (yanirs) [inline packages/explat-client/src/types.ts]: I still don't fully understand why this is needed, but I guess that being explicit is better than changing the experiment name. Maybe we can call this `requestedExperimentName`, though?

Alternatively, we can set the type of `experimentName` to `string | null`, set it to null if the requested experiment name is invalid, and drop this field. I think it's still a bit odd to be forcing the `experimentName` to something that may one day be an actual experiment name, just to return a "valid" ExperimentAssignment under an error condition that is clearly a bug that should be caught in development (invalid experiment name).

Comment (jessie-ross) [reply packages/explat-client/src/types.ts]: Hmm, yeah it's mainly from an architectural perspective I'd like all `experimentAssignments` (in general all data on the inside of a system) to be valid and consistent - even though this is a corner case. This means we don't have to have multiple `isExperimentAssignmentValid`s, ensures no future corner cases are created, and in general data is easier to reason about. For example another way to go about this might be to loosen the type to accept all strings for `experimentName`.

> Maybe we can call this requestedExperimentName, though?

👍

Comment (yanirs) [reply packages/explat-client/src/types.ts]: > For example another way to go about this might be to loosen the type to accept all strings for `experimentName`.

Hmm... That might work better, as it'd also make it easier to change the rules around names, if we ever choose to do that.

If we do go with `requestedExperimentName`, it makes sense for it to always be non-null since _something_ was always requested.

Comment (jessie-ross) [reply packages/explat-client/src/types.ts]: > Hmm... That might work better, as it'd also make it easier to change the rules around names, if we ever choose to do that.

I'm going with this, it simplifies things a lot 😄

File: packages/explat-client/src/internal/experiment-assignments.ts
Comment (yanirs) [inline packages/explat-client/src/internal/experiment-assignments.ts:36]: Now we don't need to change this line. :slightly_smiling_face:

**Review by yanirs — APPROVED** [review]
:ship:

**Comment by yanirs** [comment]
Oh, should fix the `Check code style for branches (Calypso) — TeamCity build failed` errors (Prettier issues) before merging. 

It's pretty annoying that it's "normal" for some of the build checks to fail...

**Comment by jessie-ross** [comment]
Sorry for the pings everyone!

**Pull request merged**
